Assets/HoloToolkit/ARCA/Scripts/Networking/ARCANetworkDiscovery.cs

```diff
@@ -0,0 +1,187 @@
+        [Tooltip("Component that manages the procedure of discovering new devices (mobile)")]
+        public NewDeviceDiscovery NewDeviceDiscovery;
+
+        /// <summary>
+        /// Called when the phone finds a hololens session with its marker code.
+        /// </summary>
+        public HololensSessionFoundEvent OnHololensSessionFound;
+
+        private void Awake()
+        {
+            isHost = FindObjectOfType<PlatformSwitcher>().TargetPlatform == PlatformSwitcher.Platform.Hololens;
+            //The client doesn't have to wait for the server to be started. Just give it a couple of seconds and then start it
+            if (!isHost)
+            {
+                Invoke("ManualStart", 3f);
+            }
+        }
+
+        /// <summary>
+        /// Starts the system in server or client mode depending on isHost
+        /// </summary>
+        public void ManualStart()
+        {
+            //Auto find components if necessary
+            if (MarkerGeneration3D == null)
+            {
+                MarkerGeneration3D = FindObjectOfType<MarkerGeneration3D>();
+            }
+            if (MarkerDetectionHololens == null)
+            {
+                MarkerDetectionHololens = FindObjectOfType<MarkerDetectionHololens>();
+            }
+            if (NewDeviceDiscovery == null)
+            {
+                NewDeviceDiscovery = FindObjectOfType<NewDeviceDiscovery>();
+            }
+
+            //If it isn't supposed to start listening/broadcasting exit the method
+            if (!AutoStart) return;
+
+            Initialize();
+            if (!isHost)
+            {
+                StartAsClient();
+            }
+            else
+            {
+                StartAsServer();
+                if (MarkerDetectionHololens != null)
+                {
+                    MarkerDetectionHololens.OnMarkerDetected += OnMarkerDetected;
+                }
+            }
+        }
+
+
+        #region Host
+
+        /// <summary>
+        ///     Called on the server when a marker is detected. This will update the broadcast data
+        /// </summary>
+        /// <param name="markerId"></param>
+        /// <param name="pos"></param>
+        /// <param name="rot"></param>
+        private void OnMarkerDetected( int markerId, Vector3 pos, Quaternion rot )
+        {
+            var newData = "|" + markerId + "|";
+            StartCoroutine(ChangeBroadcastData(newData));
+        }
+
+        /// <summary>
+        ///     Runs in a routine since the stop and start of the broadcasting isn't synchronous
+        /// </summary>
+        /// <param name="newData"></param>
+        /// <returns></returns>
+        private IEnumerator ChangeBroadcastData( string newData )
+        {
+            if (newData != broadcastData && !IsStopping)
+            {
+                IsStopping = true;
+                StopBroadcast();
+                yield return null; //Wait one frame for the broadcast to be stopped
```

Comment:
We've done some investigation around this and tried out different approaches: 

- Using a look to ask for ```NetworkTransport.IsBroadcastDiscoveryRunning```: It didn't work for us since there's a second Discovery running. The code used to lock itself in an infinite loop.
- Shutting down (and init) the underlying Network transport with ```NetworkTransport.Shutdown();```: It also shuts down the other components that have to keep running
- Using a timer and trust it stops/initializes: It works, but we were just guessing the time
- Current solution: **so far** it has worked every time. 

So given the architecture un Unity's networking and ARCA, we decided this solution was good enough. 

How would you do it? We can change it.

Reply:
Is there a simple check (without a lock) that you can ask if a specific broadcaster has stopped? If so, ask and yield each frame until it has stopped. If there isn't, why wait at all?

I worry about overall robustness of assuming that one frame (16.6ms @ 60fps) is enough of a wait to ensure something is shut down. I have seen too many framerate based timings fail when moved to different systems (ex: what happens if this is run on a WMR Ultra configuration where the frame time is 11.1ms?)

Reply:
Hi Dave,

I'm not aware of a way of telling whether a specific broadcasting component is running. The only thing that I found works on the NetworkTransport level and that includes all of the broadcasts. If there's a way I haven't found it :/ 

As of why to wait? Basically, if we don't wait, the component isn't stopped (or initialized) at that time. But it doesn't matter how much we wait it's always on in the next frame. And that is on the Hololens at 50-60fps, iPhone at (I guess) 60fps and in the editor at a bazillion fps(3-4ms frame time). 

**This is a guess**
*I don't know exactly why my guess is that because it works at different framerates it's not the time that matters but the fact that it advances a frame. Possibly the ```StopBroadcast``` and ```StartBroadcast``` queue a message that gets processed at the end of each frame by the Network engine?*

That would explain why, but **I have no proof** and the online documentation isn't great in that aspect. 

I know it doesn't sound very robust, I have a hard time myself defending that code, If anyone has an idea on how to do it properly I would be uber happy to change it.